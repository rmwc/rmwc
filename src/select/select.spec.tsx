import React from 'react';
import { Select } from './';
import { render, act, fireEvent, screen } from '@testing-library/react';

test('renders learn react link', (done) => {
  const onChange = jest.fn();
  const { container } = render(
    <Select label="myLabel" onChange={onChange} options={['cookie', 'pizza']} />
  );

  const select = container.children[0].querySelector('select');

  fireEvent.change(container);

  act(() => {
    window.requestAnimationFrame(() => {
      fireEvent.change(select!);
      expect(onChange).toHaveBeenCalled();
      done();
    });
  });
});

describe('Select', () => {
  it('renders', () => {
    const { asFragment } = render(
      <Select
        placeholder="Select a food"
        options={{ 1: 'Cookies', 2: 'Pizza', 3: 'Icecream' }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('helpText', () => {
    const el1 = render(
      <div>
        <Select
          helpText="selectHelpText1"
          options={{ 1: 'Cookies', 2: 'Pizza', 3: 'Icecream' }}
        />
      </div>
    );

    const el2 = render(
      <div>
        <Select
          helpText={{
            children: 'selectHelpText2',
            validationMsg: true,
            persistent: true
          }}
          options={{ 1: 'Cookies', 2: 'Pizza', 3: 'Icecream' }}
        />
      </div>
    );

    expect(screen.getByText('selectHelpText1')).toBeInTheDocument();
    expect(screen.getByText('selectHelpText2')).toBeInTheDocument();

    expect(
      el2.container.getElementsByClassName(
        'mdc-select-helper-text--validation-msg'
      )
    ).toHaveLength(1);
    expect(
      el2.container.getElementsByClassName('mdc-select-helper-text--persistent')
    ).toHaveLength(1);
  });

  it('can have empty placeholder', () => {
    const { container } = render(
      <Select
        placeholder=""
        options={{ 1: 'Cookies', 2: 'Pizza', 3: 'Icecream' }}
      />
    );

    // make sure the label is not floating
    expect(
      container.getElementsByClassName('mdc-select__label--float-above')
    ).toHaveLength(0);

    // make sure we have 4 options
    expect(screen.getAllByRole('option')).toHaveLength(4);
  });

  it('can accept options array', () => {
    const { asFragment } = render(
      <Select placeholder="Select a food" options={['1', '2', '3']} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be outlined', () => {
    const { container } = render(<Select outlined options={['1', '2', '3']} />);
    expect(container.firstChild).toHaveClass('mdc-select--outlined');
  });

  it('can accept formatted options array', () => {
    const { asFragment } = render(
      <Select
        placeholder="Select a food"
        options={[
          {
            label: 'Cookies',
            value: '1'
          },
          {
            label: 'Pizza',
            value: '2',
            'aria-disabled': true,
            tabIndex: -1
          },
          {
            label: 'Icecream',
            value: '3'
          }
        ]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('can have a tab index', () => {
    const { container } = render(
      <Select tabIndex={1} options={['1', '2', '3']} />
    );
    expect(
      container.getElementsByClassName('rmwc-select__native-control')[0]
    ).toHaveAttribute('tabindex', '1');
  });

  it('can have custom rootProps', () => {
    const { asFragment } = render(<Select rootProps={{ name: 'test' }} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can be disabled', () => {
    const selectInput = render(
      <Select disabled={false} options={['1', '2', '3']} />
    );

    expect(
      selectInput.container.getElementsByClassName('mdc-select--disabled')
    ).toHaveLength(0);

    selectInput.rerender(<Select disabled={true} options={['1', '2', '3']} />);

    expect(
      selectInput.container.getElementsByClassName('mdc-select--disabled')
    ).toHaveLength(1);
  });

  it('can have custom classnames', () => {
    const { container } = render(
      <Select
        className={'my-custom-classname'}
        options={{ 1: 'Cookies', 2: 'Pizza', 3: 'Icecream' }}
      />
    );
    expect(container.getElementsByClassName('mdc-select')[0]).toHaveClass(
      'my-custom-classname'
    );
  });

  it('can autofocus', () => {
    const { container } = render(
      <Select options={['one', 'two', 'three']} autoFocus />
    );
    expect(document.activeElement).toBe(container.querySelector('select'));
  });
});

describe('Select: Lifecycle', () => {
  it('SelectedText is blank with no value', () => {
    const { container } = render(
      <Select options={['Cookies', 'Pizza', 'Icecream']} />
    );
    expect(
      container.getElementsByClassName('mdc-select__selected_text')
    ).toHaveLength(0);
  });

  it('SelectedText is blank with no value, enhanced', () => {
    const { container } = render(
      <Select enhanced options={['Cookies', 'Pizza', 'Icecream']} />
    );
    expect(
      container.getElementsByClassName('mdc-select__selected_text')
    ).toHaveLength(0);
  });

  it('SelectedText is blank with incorrect defaultValue', () => {
    const { container } = render(
      <Select options={['Cookies', 'Pizza', 'Icecream']} defaultValue="Foo" />
    );
    expect(
      container.getElementsByClassName('mdc-select__selected_text')
    ).toHaveLength(0);
  });

  it('SelectedText is blank with incorrect defaultValue, enhanced', () => {
    const { container } = render(
      <Select
        enhanced
        options={['Cookies', 'Pizza', 'Icecream']}
        defaultValue="Foo"
      />
    );
    expect(
      container.getElementsByClassName('mdc-select__selected_text')
    ).toHaveLength(0);
  });

  it('SelectedText is set to value', () => {
    render(
      <Select options={['Cookies', 'Pizza', 'Icecream']} value="Cookies" />
    );
    expect(screen.getAllByText('Cookies')[0]).toHaveClass(
      'mdc-select__selected-text'
    );
  });

  it('SelectedText is set to value, enhanced', () => {
    render(
      <Select
        enhanced
        options={['Cookies', 'Pizza', 'Icecream']}
        value="Cookies"
      />
    );
    expect(screen.getAllByText('Cookies')[0]).toHaveClass(
      'mdc-select__selected-text'
    );
  });

  it('SelectedText is set to default value', () => {
    render(
      <Select
        options={['Cookies', 'Pizza', 'Icecream']}
        defaultValue="Cookies"
      />
    );
    expect(screen.getAllByText('Cookies')[0]).toHaveClass(
      'mdc-select__selected-text'
    );
  });

  it('SelectedText is set to default value, enhanced', () => {
    render(
      <Select
        enhanced
        options={['Cookies', 'Pizza', 'Icecream']}
        defaultValue="Cookies"
      />
    );
    expect(screen.getAllByText('Cookies')[0]).toHaveClass(
      'mdc-select__selected-text'
    );
  });
});
