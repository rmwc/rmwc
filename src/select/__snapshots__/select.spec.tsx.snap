// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select can accept formatted options array 1`] = `
<DocumentFragment>
  <div
    class="mdc-select mdc-select--filled mdc-select--no-label"
  >
    <div
      aria-disabled="false"
      aria-haspopup="listbox"
      class="mdc-select__anchor"
      role="button"
      tabindex="0"
    >
      <span
        class="mdc-line-ripple mdc-line-ripple--deactivating"
        style="transform-origin: 0px center;"
      />
      <span
        class="mdc-floating-label"
        for="select-test"
      />
      <span
        class="mdc-select__selected-text-container"
      >
        <span
          class="mdc-select__selected-text"
        >
          Select a food
        </span>
      </span>
      <span
        class="mdc-select__dropdown-icon"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-inactive"
        >
          arrow_drop_down
        </i>
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-active"
        >
          arrow_drop_up
        </i>
      </span>
      <select
        class="rmwc-select__native-control "
        tabindex="0"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select a food
        </option>
        <option
          value="1"
        >
          Cookies
        </option>
        <option
          aria-disabled="true"
          tabindex="-1"
          value="2"
        >
          Pizza
        </option>
        <option
          value="3"
        >
          Icecream
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Select can accept options array 1`] = `
<DocumentFragment>
  <div
    class="mdc-select mdc-select--filled mdc-select--no-label"
  >
    <div
      aria-disabled="false"
      aria-haspopup="listbox"
      class="mdc-select__anchor"
      role="button"
      tabindex="0"
    >
      <span
        class="mdc-line-ripple mdc-line-ripple--deactivating"
        style="transform-origin: 0px center;"
      />
      <span
        class="mdc-floating-label"
        for="select-test"
      />
      <span
        class="mdc-select__selected-text-container"
      >
        <span
          class="mdc-select__selected-text"
        >
          Select a food
        </span>
      </span>
      <span
        class="mdc-select__dropdown-icon"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-inactive"
        >
          arrow_drop_down
        </i>
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-active"
        >
          arrow_drop_up
        </i>
      </span>
      <select
        class="rmwc-select__native-control "
        tabindex="0"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select a food
        </option>
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Select can have custom rootProps 1`] = `
<DocumentFragment>
  <div
    class="mdc-select mdc-select--filled mdc-select--no-label"
    name="test"
  >
    <div
      aria-disabled="false"
      aria-haspopup="listbox"
      class="mdc-select__anchor"
      role="button"
      tabindex="0"
    >
      <span
        class="mdc-line-ripple mdc-line-ripple--deactivating"
        style="transform-origin: 0px center;"
      />
      <span
        class="mdc-floating-label"
        for="select-test"
      />
      <span
        class="mdc-select__selected-text-container"
      >
        <span
          class="mdc-select__selected-text"
        />
      </span>
      <span
        class="mdc-select__dropdown-icon"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-inactive"
        >
          arrow_drop_down
        </i>
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-active"
        >
          arrow_drop_up
        </i>
      </span>
      <select
        class="rmwc-select__native-control "
        tabindex="0"
      >
        <option
          disabled=""
          selected=""
          value=""
        />
      </select>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Select renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-select mdc-select--filled mdc-select--no-label"
  >
    <div
      aria-disabled="false"
      aria-haspopup="listbox"
      class="mdc-select__anchor"
      role="button"
      tabindex="0"
    >
      <span
        class="mdc-line-ripple mdc-line-ripple--deactivating"
        style="transform-origin: 0px center;"
      />
      <span
        class="mdc-floating-label"
        for="select-test"
      />
      <span
        class="mdc-select__selected-text-container"
      >
        <span
          class="mdc-select__selected-text"
        >
          Select a food
        </span>
      </span>
      <span
        class="mdc-select__dropdown-icon"
      >
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-inactive"
        >
          arrow_drop_down
        </i>
        <i
          class="rmwc-icon rmwc-icon--ligature material-icons mdc-select__dropdown-icon-active"
        >
          arrow_drop_up
        </i>
      </span>
      <select
        class="rmwc-select__native-control "
        tabindex="0"
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          Select a food
        </option>
        <option
          value="1"
        >
          Cookies
        </option>
        <option
          value="2"
        >
          Pizza
        </option>
        <option
          value="3"
        >
          Icecream
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
